py/dml/c_backend.py

```diff
@@ -880,36 +874,78 @@ def port_class_ident(port):
     initialization function'''
     return '_port_class_' + port.attrname()
 
+def port_prefix(port):
+    if port.objtype == 'port':
+        return '' if param_bool(port, '_is_subdevice_port') else 'port.'
+    else:
+        assert port.objtype == 'bank'
+        return 'bank.'
+
+def find_port_parent(port, indices):
+    '''Given a port or bank node, find the port node that correspond to
+    port object's port parent, and the name of the port object
+    relative to port parent and the device. The return value is a
+    triple (node, prefix, suffix), where node is the port parent,
+    suffix is the conf-object name as descendant of the port parent,
+    and (prefix+suffix) is the conf-object name as descendant of the
+    device. E.g., for a node subdev.p, returns (subdev.obj, "subdev",
+    "port.p").'''
+    ancestor = port.parent
+    suffix_parts = [port_prefix(port) + port.name_anonymized
+                    + ''.join(f'[{i}]' for i in indices[ancestor.dimensions:])]
+    indices = indices[:ancestor.dimensions]
+    while ancestor.objtype not in {'device', 'port'}:
+        if ancestor.objtype == 'port':
+            assert param_bool(ancestor, '_is_subdevice_port')
+        else:
+            assert ancestor.objtype == 'group'
+        name = ancestor.name_anonymized
+        idx = ''.join(f'[{i}]' for i in indices[ancestor.parent.dimensions:])
+        suffix_parts.append(f'{name}{idx}.')
+        ancestor = ancestor.parent
+        indices = indices[:ancestor.dimensions]
+    if ancestor.objtype == 'device':
+        prefix = ''
+    else:
+        (_, prefix_prefix, prefix_suffix) = find_port_parent(ancestor, indices)
+        prefix = f'{prefix_prefix}{prefix_suffix}.'
+    return (ancestor, prefix, ''.join(reversed(suffix_parts)))
+
+def node_ancestors(node, since):
+    while True:
+        if node == since:
```

Comment:
This is an identity check, yeah? If so, `is` instead of `==`.

Reply:
done